### PR TITLE
Implement event source feature with file get and watch source

### PR DESCRIPTION
**Pull request by william-gr**
Ticket: #26278

### CODE REVIEW

**Comment by william-gr** [comment]
@themylogin I am not entirely sure of this implementation, maybe it could be done in a better way but it seems to work. Can you review and let me know what you think? Thanks!

File: src/middlewared/middlewared/plugins/filesystem.py [hunk 1]
```diff
@@ -138,3 +140,57 @@ def file_get_contents(self, path, options=None):
                 f.seek(options['offset'])
             data = binascii.b2a_base64(f.read(options.get('maxlen'))).decode().strip()
         return data
+
+
+class GetAndWatchEventSource(EventSource):
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/filesystem.py]: I wonder if less generic and more clear name exists? Like `FollowTailEventSource` (like in `tail -f`)

File: src/middlewared/middlewared/plugins/filesystem.py [hunk 2]
```diff
@@ -138,3 +140,57 @@ def file_get_contents(self, path, options=None):
...
+            return
+
+        bufsize = 8192
+        fsize = os.stat(path).st_size
+        if fsize < bufsize:
+            bufsize = fsize
+        i = 0
+        with open(path) as f:
+            data = []
+            while True:
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/filesystem.py]: More efficient implementation is possible if you are really into this :)
For example, this is how it's done in GNU tail: http://git.savannah.gnu.org/cgit/coreutils.git/tree/src/tail.c#n502

File: src/middlewared/middlewared/plugins/filesystem.py [hunk 3]
```diff
@@ -138,3 +140,57 @@ def file_get_contents(self, path, options=None):
...
+                data.extend(f.readlines())
+                if len(data) >= lines or f.tell() == 0:
+                    break
+
+            self.send_event('ADDED', fields={'data': ''.join(data[-lines:])})
+            f.seek(fsize)
+
+            kqueue = select.kqueue()
+
+            ev = [select.kevent(
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/filesystem.py]: I wonder if we can just abandon this and call `tail` through `subprocess` :)

That way we can also get `-F` for free

File: src/middlewared/middlewared/client/client.py [hunk 4]
```diff
@@ -545,6 +548,18 @@ def from_json(args):
                         else:
                             data.append(str(f))
                     print('|'.join(data))
+
+    elif args.name == 'subscribe':
+        with Client(uri=args.uri) as c:
+
+            event = TEvent()
```
[the remark this answers is not in the document]
Comment (william-gr) [reply src/middlewared/middlewared/client/client.py]: Good question!

File: src/middlewared/middlewared/client/client.py [hunk 5]
```diff
@@ -545,6 +548,18 @@ def from_json(args):
...
+
+            def cb(mtype, **message):
+                print(mtype, message)
+
+            c.subscribe(args.event, cb)
+
+            event.wait()
```
[the remark this answers is not in the document]
Comment (william-gr) [reply src/middlewared/middlewared/client/client.py]: So far nobody, this is just for development purposes. Perhaps I will add "wait for X events" then set it, in the future.

File: src/middlewared/middlewared/plugins/filesystem.py (hunk 1 above)
Comment (william-gr) [reply src/middlewared/middlewared/plugins/filesystem.py]: That indeed sounds better

File: src/middlewared/middlewared/plugins/filesystem.py (hunk 3 above)
Comment (william-gr) [reply src/middlewared/middlewared/plugins/filesystem.py]: That would be possible, however I am allergic to subprocess calls =P

File: src/middlewared/middlewared/main.py [hunk 6]
```diff
@@ -177,15 +180,37 @@ def send_error(self, message, errno, reason=None, exc_info=None, etype=None, ext
...
+
+        if ':' in name:
+            shortname, arg = name.split(':', 1)
+        else:
+            shortname = name
+            arg = None
+        event_source = self.middleware.get_event_source(shortname)
+        if event_source:
+            es = event_source(self.middleware, self, ident, name, arg)
+            t = threading.Thread(target=es.process, daemon=True)
```
[the remark this answers is not in the document]
Comment (william-gr) [reply src/middlewared/middlewared/main.py]: Helper function to what exactly? threading.Thread?

File: src/middlewared/middlewared/main.py [hunk 7]
```diff
@@ -194,6 +219,11 @@ def send_event(self, name, event_type, **kwargs):
                 found = True
                 break
         if not found:
+            for i in self.__event_sources.values():
+                if i['name'] == name:
+                    found = True
```
[the remark this answers is not in the document]
Comment (william-gr) [reply src/middlewared/middlewared/main.py]: This has the disadvantage of looping through all the items in the dict, even it matches at first.

Not sure how for...else would improve this, can you elaborate a bit further?

**Comment by william-gr** [comment]
@themylogin thanks for the review! I have addressed a couple of them and followed up on the other questions ;)

File: src/middlewared/middlewared/main.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (themylogin) [reply src/middlewared/middlewared/main.py]: > This has the disadvantage of looping through all the items in the dict, even it matches at first.

No

https://docs.python.org/3/library/functions.html#any

```
>>> def f(x):
...     print(x)
...     return True if x % 2 else False
... 
>>> any(f(x) for x in [0,1,2,3])
0
1
True
>>> 
```

> Not sure how for...else would improve this, can you elaborate a bit further?

```
for i in self.__event_sources.values():
    if i['name'] == name:
        break
else:
    return
```

File: src/middlewared/middlewared/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (themylogin) [reply src/middlewared/middlewared/main.py]: I have `start_daemon_thread(target)` in most of my projects. I also catch (and log / send to `sentry`) thread body exceptions there.

**Comment by william-gr** [comment]
@themylogin thnaks for the `any` tip, it indeed looks nicer.

I have committed a couple more concerns, let me know if there is anything left.

**Pull request merged**